design/007-external-data-handling/007-external-data-handling.md

```diff
@@ -0,0 +1,601 @@
+that can be turned off in more restrictive environments. The alternative that
+will be provided is to mount volumes at the start of the user session, requiring
+a restart if a new volume is needed. That way, an entirely separate service can
+handle the volume provisioining, as is the case now with the current
+implementation of S3 buckets.
+
+
+### Commands
+
+Below are some examples of potential usage of external data with existing and
+new `renku` commands.
+
+There are a few situations to consider. In all cases the data referenced from
+external storage should be trackable through renku workflows and the Renku API
+should offer utilities to access that data.
+
+1. External (raw) data exists in some external location and it would be useful
+   to bring it to Renku and annotate it in the context of a Renku dataset. The
+   difference with current functionality is that the user does not want to add
+   this data to git-LFS.
+
+2. A Renku dataset is created but instead of using git-LFS it should be backed
+   by another external data store (perhaps for easier access from other
+   systems). This means that _local_ data will be added to the dataset and
+   (eventually) sent to the remote storage.
+
+3. External data is brought into the project but _not_ in the context of a
+   dataset (that could be done at a later point). Renku should remember that
+   this data has been brought into the project so that other users of the
+   project can also easily stage it when needed. As above, the data should _not_
+   be pushed to git-LFS but instead kept on the remote resource.
+
+In all three of these cases, at least two of the access modes described in the
+previous section should be accommodated, namely "copy" and "mount". In the case
+of datasets, it's not clear how "reference" is to be used; however, it doesn't
+seem contradictory to organize data references in a dataset and make them
+accessible via Renku's `Dataset` API.
+
+Note: in the commands below `s3://` is used as an example but all of the support
+libraries referred to above support common APIs for (nearly) all likely storage
+backends and automatically switch between them depending on the protocol of the
+URL. That's a long way of saying that one can replace `s3://` with `hdfs://`,
+`nfs://` or `gcs://` etc.
+
+#### Creating a Renku dataset backed by external storage
+
+```
+renku dataset create s3-backed-dataset --storage s3://server/path/to/bucket
+```
+
+This command creates a dataset and sets its backing storage to be S3 instead of
+the default git-LFS. Locally, the directory `data/s3-backed-dataset` is created
+and added to `.gitignore`.
+
+```
+renku dataset add s3-backed-dataset /path/to/file/on/filesystem
+```
+
+This adds a file from the local filesystem to the dataset. It is _copied_ to
+`data/s3-backed-dataset`. (Should it also be pushed to the remote storage
+immediately?)
+
+#### Adding external data to a Renku Dataset
```

Comment:
Should we allow adding data from an external storage to an already-existing dataset? In other words, can a dataset contain data from multiple external sources? I don't see any reason not to do that. However, in this case we need to specify a data source (to push data) when we add local data to a dataset (e.g. an output of a computation).

Reply:
This is a good point. 

> However, in this case we need to specify a data source (to push data)

You mean if a dataset has multiple sources defined, then on push you would have to say which destination should receive the newly added files?